R/method-correlation.R

```diff
@@ -59,9 +59,205 @@ setMethod("correlation", signature("data.table", "data.table"), function(data1,
+#' 
+#' @param data1 data.table with columns as variables. All columns must be numeric. One row per sample.
+#' @param method string defining the type of correlation to run. The currently supported values are 'spearman' and 'pearson'
+#' @param verbose boolean indicating if timed logging is desired
+#' @return data.frame with correlation coefficients
+setMethod("correlation", signature("data.table", "missing"), function(data1, data2, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) {
+  # Check that all values are numeric
+  if (!identical(veupathUtils::findNumericCols(data1), names(data1))) { stop("All columns in data1 must be numeric.")}
+
+  ## Compute correlation
+  # rownames and colnames should be the same in this case
+  # na.or.complete removes rows with NAs, if no rows remain then correlation is NA
+  corrResult <- data.table::as.data.table(cor(data1, method = method, use='na.or.complete'), keep.rownames = T)
+  veupathUtils::logWithTime(paste0('Completed correlation with method=', method,'. Formatting results.'), verbose)
+
+  ## Format results
+  meltedCorrResult <- melt(corrResult, id.vars=c('rn'))
+  formattedCorrResult <- data.frame(
+    data1 = meltedCorrResult[['rn']],
+    data2 = meltedCorrResult[['variable']],
+    correlationCoef = meltedCorrResult[['value']]
+  )
+
+  return(formattedCorrResult)
+})
+
+getDataMetadataType <- function(data) {
+  if (inherits(data, 'AbundanceData')) {
+    return('assay')
+  } else if (inherits(data, 'SampleMetadata')) {
+    return('sampleMetadata')
+  } else {
+    stop('data must be an AbundanceData or SampleMetadata object')
+  }
+}
+
+## Helper function
+# should this be s4?
+buildCorrelationComputeResult <- function(corrResult, data1, data2 = NULL, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) {
+  method <- veupathUtils::matchArg(method)
+  verbose <- veupathUtils::matchArg(verbose)
+  if (!inherits(data1, c("AbundanceData", "SampleMetadata"))) {
+    stop('data1 must be an AbundanceData or SampleMetadata object')
+  }
+
+  # both AbundanceData and SampleMetadata have these slots
+  recordIdColumn <- data1@recordIdColumn
+  ancestorIdColumns <- data1@ancestorIdColumns
+  allIdColumns <- c(recordIdColumn, ancestorIdColumns)
+
+  ## Format results
+  # Construct the ComputeResult
+  result <- new("ComputeResult")
+  result@name <- 'correlation'
+  result@recordIdColumn <- recordIdColumn
+  result@ancestorIdColumns <- ancestorIdColumns
+  statistics <- CorrelationResult(
+    statistics = corrResult,
+    data1Metadata = getDataMetadataType(data1),
+    data2Metadata = ifelse(is.null(data2), getDataMetadataType(data1), getDataMetadataType(data2))
+  )
+  result@statistics <- statistics
+  result@parameters <- paste0('method = ', method)
+
+  # The resulting data should contain only the samples actually used.
+  # this seems slightly complicated to generalize, and im not sure what we use it for anyhow
```

Comment:
The goal is to keep track of which columns actually got used in the computation. Sometimes we might kick columns out because of missingness or all 0s or blah. We don't have checks like that yet in correlation, but we do in other apps iirc. Anyways so the code here isn't doing anything yet, it was just left as a reminder

Reply:
ok. im a little inclined to leave it out for now and figure how to do this once we actually have a use in front of us.